### DISCUSSION TITLE
RESTful_command template for Home Assistant

### DISCUSSION THREAD

**Discussion by Djelibeybi** [delfick/photons#71] [Show and tell]
Hey folks,

I thought I'd share a `restful_command` template you can add to your Home Assistant `configuration.yaml` to simplify making a kelvin-based transformations when using the [Photons Interactor](https://github.com/delfick/photons-homeassistant) add-on. 

Essentially,  using a template means you don't have to have lots of the same JSON repeated over and over in your configuration. Instead, you can use this  simple YAML defintion.

```yaml
service: rest_command.lifx_kelvin_transform
data:
  serials:
    - d073d5000000
    - d073d51111111
  brightness: 0.9
  kelvin: 4000
  duration: 10
```

Here's what you need to add to `configuration.yaml` (or wherever you put your `restful_command` configuration!):

```yaml
rest_command:
  lifx_kelvin_transform:
    url: "http://localhost:6100/v1/lifx/command"
    method: put
    content_type: application/json
    payload: |
      {
        "command": "transform",
        "args": {
          {% if serials %}
          "matcher": {
            "serial": [
              {% for serial in serials %}
              {% if loop.last %}
              "{{ serial }}"
              {% else %}
              "{{ serial }}",
              {% endif %}
              {% endfor %}
            ]
          },
          {% endif %}
          "transform": {
            {% if brightness is defined and brightness|float and brightness >= 0 and brightness <= 1 %}
            "brightness": {{ brightness }},
            {% endif %}
            {% if kelvin is defined and kelvin|int and kelvin >= 1500 and kelvin <= 9500 %}
            "kelvin": {{ kelvin }},
            {% endif %}
            {% if duration is defined and duration|int and duration > 0 %}
            "duration": {{ duration }},
            {% endif %}
            {% if power is defined and power in ("on", "off") %}
            "power": {{ power }},
            {% endif %}
            "hue": 0,
            "saturation": 0
          },
          "transform_options": {
            "transition_color": true
          }
        }
      }
```

**Comment by joshua-pe** [comment]
Ooo, that looks great. I will have to try it out. Thanks!

**Comment by delfick** [comment]
shiny

**Comment by pimw1** [comment]
Hi, i've adopted a similar approach for applying scenes. Two improvements:

- Home Assistant terminology and units are used: "transition" instead of "duration", "brightness_pct" in the range [0:100] instead of "brightness" in the range of [0:1]
- a default value (0) is provided for the integers and floats, in line with the new best practice for templating as introduced in Home Assistant 2021.10. This helps in preventing an error when providing an empty value instead of no value.

```
rest_command:
  lifx_scene_apply:
    url: 'http://127.0.0.1:6100/v1/lifx/command'
    method: "put"
    payload: |
      {
        "command": "scene_apply",
        "args": {
          "uuid":"{{scene}}",
          "overrides": {
            {% if brightness_pct is defined and brightness_pct|float(0) and brightness_pct >= 0 and brightness_pct <= 100 %}
            "brightness": {{ brightness_pct/100 }},
            {% endif %}
            {% if kelvin is defined and kelvin|int(0) and kelvin >= 1500 and kelvin <= 9500 %}
            "kelvin": {{ kelvin }},
            {% endif %}
            {% if transition is defined and transition|int(0) and transition > 0 %}
            "duration": {{ transition }}
            {% endif %}
          }
        }
      }
```

![image](https://user-images.githubusercontent.com/74881023/136714502-54e99678-7d1b-4ae1-bbcd-bb0d82130d71.png)

**Reply by Djelibeybi** [reply]
Heh, I'd been noodling with similar changes myself. I still need to find a good way to map Home Assistant entity IDs to LIFX serials. 

I like the use of `scene_apply` instead of `transform` too. I hadn't thought of that.

**Reply by pimw1** [reply]
Thanks! Your solution was a very good basis to continue upon. Regarding the linkage of "entity id" - "lifx serial" : i have no idea, since i cannot find any attribute in home assistant capturing the lifx-serial. An ugly way to get around this, could perhaps be to manually store a reference file in the home assistant folder containing a dictonary {"entity id": "lifx serial"}

**Reply by Djelibeybi** [reply]
I was going to play with setting the entity ID == serial and use friendly name for dashboards, to see if that worked well enough.

**Reply by pimw1** [reply]
That would not be my preference, reason being that i have incorporated standard naming conventions in Home Assistant for both entity_id and friendly_name. Those naming conventions are very helpful to me.

**Reply by Djelibeybi** [reply]
Oh, not my preference either. If I can find some spare time, I'll see if I can add serial as an attribute to the LIFX integration and if upstream will accept it.

**Reply by pimw1** [reply]
That would be so nice!

**Reply by Djelibeybi** [reply]
In the meantime, I logged a bug upstream: https://github.com/home-assistant/core/issues/57457

**Reply by Djelibeybi** [reply]
And just submitted a PR that fixes the bug: https://github.com/home-assistant/core/pull/58457 

It's my first PR to Home Assistant, so I have no idea if they'll merge it or not.

**Comment by pimw1** [comment]
While thinking about it: the serial equals the mac address of the device. With an unifi access point, you can use the unifi integration to find the mac address. Could be something?
![image](https://user-images.githubusercontent.com/74881023/136715036-b3c61805-81b9-4575-86bb-c241730818f5.png)

**Reply by Djelibeybi** [reply]
Except the serial no longer equals the MAC address, at least not for Gen3/Gen4 bulbs. For some reason, the latest firmware incremented the MAC address by one.

**Reply by delfick** [reply]
this is correct. Annoying reality of a particular feature those lights have.

**Comment by pimw1** [comment]
I've expanded the functionality a little bit.

- It is now possible to provide a kelvin value from any choosen switch from the custom component adaptive lighting ( https://github.com/basnijholt/adaptive-lighting ) . Manual entering of a kelvin value is still possible of course.
- The code is a little bit more robust (less prone to user errors) and more efficient

```
rest_command:
  lifx_scene_apply:
    url: 'http://127.0.0.1:6100/v1/lifx/command'
    method: "put"
    payload: |
      {
        "command": "scene_apply",
        "args": {
          "uuid":"{{scene}}",
          "overrides": {
            {% if brightness_pct is defined and is_number(brightness_pct) == True and 0 <= brightness_pct|float(0) <= 100 %}
                "brightness": {{ brightness_pct/100 }},
            {% else %}
                "dummy_brightness": 0,
            {% endif %}
            {% if kelvin is defined and is_number(kelvin) == True and 1500 <= kelvin|int(0) <= 9500 %}
                "kelvin": {{ kelvin }},
            {% elif kelvin is defined  and is_number(kelvin) == False and 1500 <= state_attr(kelvin, 'color_temp_kelvin')|int(0) <= 9500 %}
                "kelvin": {{state_attr(kelvin, 'color_temp_kelvin')}},
            {% else %}
                "dummy_kelvin": 0,
            {% endif %}
            {% if transition is defined and is_number(transition) == True and 0 <= transition|int(0) %}
                "duration": {{ transition }}
            {% else %}
                "dummy_duration": 0
            {% endif %}
          }
        }
      }
```
![image](https://user-images.githubusercontent.com/74881023/137038230-0ddcd3c3-e05c-4bf5-80c4-c4ce225773b4.png)
->
![image](https://user-images.githubusercontent.com/74881023/137038587-cc876ced-ee9c-42ff-afc6-84c861b5dbcc.png)
->
![image](https://user-images.githubusercontent.com/74881023/137038342-5e6883a6-9c5e-4e65-9853-3afd6bfaae3a.png)


![image](https://user-images.githubusercontent.com/74881023/137038153-d2fad944-2f75-41c8-8025-c7510d4307c4.png)

**Reply by Djelibeybi** [reply]
That's awesome. I use Node-RED for actual automnations, so I was pulling the kelvin value from Adaptive Lighting in my Flow.

**Reply by pimw1** [reply]
Thanks!

By pushing this logic down towards the rest_command, everything else build on top of it becomes simpeler (node-red, automations, scripts)..... or at least i hope this will make my automations somewhat simpeler haha.

**Reply by Djelibeybi** [reply]
Oh, I agree. It also ensures that whatever you chose to use to automate, the outcome when using a specific rest_command is the same. I've added this to my "should apply" list. :)

**Comment by Djelibeybi** [comment]
Here's another RESTful Command, this time to trigger a two-hour clean cycle on all available HEV bulbs:

```yaml
lifx_clean:
  url: "http://localhost:6100/v1/lifx/command"
  method: put
  content_type: application/json
  payload: |
    {
      "command": "clean/start",
      "args": {
        "matcher": "cap=hev",
        "duration": 7200
      }
    }
```
